### PR TITLE
refactor(ir): remove DestructValue expressions

### PR DESCRIPTION
**Pull request by kszucs**
Defer `DestructValue` expression's functionality to`ops.StructField` which removes specific execution branches from pandas and dask backends.

### CODE REVIEW

File: ibis/backends/tests/test_vectorized_udf.py
Comment (kszucs) [inline ibis/backends/tests/test_vectorized_udf.py:519]: Since this PR changes the IR the execution model changes as well, which means that currently we cannot reuse the results coming from the same UDF call for separate operations.

I think this should be rather implemented in the execution model of pandas since we should be able to reuse results from different branches of the expression tree.

**Comment by cpcloud** [comment]
@jreback @gerrymanoim @emilyreff Does this affect y'all? This PR provides an alternative to the existing destructuring API, but it's a breaking change.

**Review by cpcloud — CHANGES_REQUESTED** [review]

File: ibis/backends/tests/test_vectorized_udf.py
Comment (kszucs) [reply ibis/backends/tests/test_vectorized_udf.py:519]: It is.

File: ibis/udf/vectorized.py
Comment (kszucs) [reply ibis/udf/vectorized.py:35]: Right, updating.

Comment (kszucs) [reply ibis/udf/vectorized.py]: Updated, though it's rather hard to test since the coercion function operates on either sequence of pandas series or mapping of pandas series.

File: ibis/backends/dask/execution/structs.py
Comment (kszucs) [reply ibis/backends/dask/execution/structs.py:16]: I'm not sure I'm following, what else should it select?

File: ibis/backends/tests/test_vectorized_udf.py
Comment (cpcloud) [reply ibis/backends/tests/test_vectorized_udf.py]: @gerrymanoim @jreback @emilyreff7 @icexelloss Wondering if y'all are affected by this.

File: ibis/backends/dask/execution/util.py
Comment (gerrymanoim) [inline ibis/backends/dask/execution/util.py:93]:
```suggestion
    Examples below use pandas objects for legibility, but functionality is the
```

**Comment by kszucs** [comment]
I feel like `DestructValue` is an optimization leaked from the pandas backend to the expression tree level.

On the long term we should refactor the pandas backend to properly reuse intermediate results (in this case the UDF result) similarly to the pyspark backend.

**Comment by icexelloss** [comment]
@cpcloud @kszucs This likely will affect us

> This PR provides an alternative to the existing destructuring API

Can you please provide high level of the new API?

**Comment by kszucs** [comment]
> @cpcloud @kszucs This likely will affect us
> 
> > This PR provides an alternative to the existing destructuring API
> 
> Can you please provide high level of the new API?

This PR doesn't change the API actually, just replaces the [destructure operation using StructField ops](https://github.com/ibis-project/ibis/pull/3497/files#diff-8f6e03b936fae42459cef36d6bb2691392fb1ed9bd03c9846152ac6ebb6da892R4037). The idea here would be to let the backend do the optimization, like PySpark is able to reuse intermediate UDF results whereas the pandas backend is not.

File: ibis/backends/tests/test_vectorized_udf.py
Comment (kszucs) [reply ibis/backends/tests/test_vectorized_udf.py:501]: Yes, that's the actual breaking change.

Comment (icexelloss) [reply ibis/backends/tests/test_vectorized_udf.py:501]: I remembered that pandas does have "exact once" semantics for destructure UDFs before. Is it that true?

Comment (kszucs) [reply ibis/backends/tests/test_vectorized_udf.py:501]: If we were about to merge this PR, then yes. The behavior could be still still restored by refactoring the pandas execution to reuse/cache results of the expressions (with additional speedups).

File: ibis/udf/vectorized.py
Comment (kszucs) [reply ibis/udf/vectorized.py:210]: Otherwise [ops.StructField implementation](https://github.com/ibis-project/ibis/pull/3497/files#diff-0eb3b553dfe84f503ecd43188a21a2df3aa106437939772d608c7c13f09722bcR15) would have to deal with both integer and string indices.

File: ibis/expr/analysis.py
Comment (icexelloss) [inline ibis/expr/analysis.py:568]: Do the new code produces the same column ordering as before? I remembered this was a pretty nasty issue before.

Comment (icexelloss) [reply ibis/expr/analysis.py:568]: (When part of the destructure columns overwrites existing columns)

File: ibis/backends/tests/test_vectorized_udf.py
Comment (icexelloss) [reply ibis/backends/tests/test_vectorized_udf.py:501]: > If we were about to merge this PR, then yes.

Just to confirm:
Before the PR - we have exact once for pandas
After the PR - we don't have exact once for pandas
?

File: ibis/expr/analysis.py
Comment (kszucs) [reply ibis/expr/analysis.py:568]: The column ordering now should be identical as calling

```python
t.mutate(
    t.struct_field['a'],
    t.struct_field['b'],
    t.struct_field['c'],
)
```

File: ibis/backends/tests/test_vectorized_udf.py
Comment (kszucs) [reply ibis/backends/tests/test_vectorized_udf.py:501]: Correct.

Comment (icexelloss) [reply ibis/backends/tests/test_vectorized_udf.py:501]: This is going to be an issue for us because internally we use this feature pretty heavily and would be a perf regression if we started to execute this types of UDF multiple times...

File: ibis/expr/analysis.py
Comment (icexelloss) [reply ibis/expr/analysis.py:568]: kk sounds good. Just double checking.

File: ibis/backends/tests/test_vectorized_udf.py
Comment (kszucs) [reply ibis/backends/tests/test_vectorized_udf.py:501]: Thanks for checking! I'll try to find a solution to keep the previous behavior then.

Comment (icexelloss) [reply ibis/backends/tests/test_vectorized_udf.py:501]: To give a bit more context, during previous work. I think we tried previous to destructure the assignment directly (similar to what you are doing) but @emilyreff noticed that Pandas backend cannot share intermediate result among different struct field object that points to the same struct and opened https://github.com/ibis-project/ibis/issues/2649

And I believe this is the reason that the code is written in the current way... (with a destructure expr)

Comment (icexelloss) [reply ibis/backends/tests/test_vectorized_udf.py:501]: (Sorry should probably have linked the issue in code comments to be clear)

**Comment by cpcloud** [comment]
@kszucs What's the status of this PR?

**Comment by kszucs** [comment]
Currently blocked on implementing support for the pandas backend to reuse intermediate results, thus cache the output of udfs returning with struct types.

**Comment by cpcloud** [comment]
@kszucs Should we move this to a draft PR for now?

**Comment by kszucs** [comment]
I cannot move it to draft. I think we can keep it open so we can merge this once the pandas backend is able to reuse intermediate results.

**Comment by cpcloud** [comment]
@kszucs Let's close this out. There's already of merge conflicts and more refactoring to do to make this work.

**Pull request closed without merging**
